### PR TITLE
Purchasable Emergency Shuttles (Ready for review)

### PR DESCRIPTION
**Pull request by lewcc**
## What Does This PR Do
Adds purchasable emergency shuttles, inspired by the feature on /tg/.

For a given amount of cargo points, the crew will be able to purchase an alternative emergency shuttle that will arrive at the end of shift, through crew transfer or emergency call. As of this PR, it includes all of the standard emergency shuttles that you might be familiar with, such as:

- The cramped shuttle
- The clown shuttle
- The retro shuttle
- The departmental shuttle

and so on and so forth, with the intention to add new shuttles as part of later PRs.
Some of these shuttles will be available normally (such as the retro one, and the departmental one), while others (such as the jungle shuttle and the nar'sie shuttle) will be available through an emagged console. 

This PR also reworks some of the shuttle code by adding the framework for modifiable shuttle speed. Some of these new shuttles might arrive a little quicker than others, or may be slower. Mind your time!

Shuttle code aside, this *also* provides the framework for abstract cargo orders, or orders that don't return anything on the shuttle and are instead just completed once paid for and approved. This is how the shuttle purchasing works, but this would also likely be useful on the admin side to allow the crew to purchase certain things in events. I've added a sample item that messages admins when it is purchased, and I'd love to look into ways for making this easier to use for admins down the line.

On the back-end, this PR also notably moves some of the shuttle loading code out of the shuttle manipulator, and into the shuttle subsystem itself. I've tested the manipulator as is, and it seems to behave the same.

## Why It's Good For The Game
Being able to purchase new escape shuttles adds quite a bit of diversity to something that's generally only modified through adminbus. It provides a great money-sink for the crew on every round, and something that everyone might be able to pool their paychecks for, or a rallying point for a crew. The ability for admins to create abstract purchases as well might also provide for neat, more hands-off ways to run events.

## Images of changes

https://github.com/ParadiseSS13/Paradise/assets/89928798/6fd5fc39-1d68-419b-98f1-34b1db0b459b

## Testing

Ran through and tested the shuttle manipulator to make sure it still behaves as it did before, and also ordered a bunch of shuttles.

This will need a fair amount of testing once completed as well.

## Changelog
:cl:
add: Alternative emergency shuttles can now be purchased from the cargo supply console.
tweak: Shuttles can now have different speed factors, adjusting how long their transit time takes.
add: The supply console can now be cmagged to order a silly shuttle.
/:cl:

### CODE REVIEW

File: code/__DEFINES/economy_defines.dm [hunk 1]
```diff
@@ -60,8 +60,25 @@
...
+	/datum/map_template/shuttle/emergency/military,
+	/datum/map_template/shuttle/emergency/clown,
+	/datum/map_template/shuttle/emergency/cramped,
+	/datum/map_template/shuttle/emergency/narnar,
+	/datum/map_template/shuttle/emergency/old,
+	/datum/map_template/shuttle/emergency/jungle,
+	/datum/map_template/shuttle/admin/skipjack  // hehe
+
+
+))
```
Comment (Sirryan2002) [inline code/__DEFINES/economy_defines.dm:84]: just put this on SSshutles or SSeconomy tbh

File: code/controllers/subsystem/SSshuttles.dm [hunk 2]
```diff
@@ -303,4 +305,99 @@ SUBSYSTEM_DEF(shuttle)
...
+	//   the shuttle.
+	// - We need to check that no additional ports have slipped in from the
+	//   template, because that causes unintended behaviour.
+	for(var/T in affected)
+		for(var/obj/docking_port/P in T)
+			if(istype(P, /obj/docking_port/mobile))
+				port = P
+				mobile_docking_ports++
+				if(mobile_docking_ports > 1)
+					qdel(P, force=TRUE)
```
Comment (Sirryan2002) [inline code/controllers/subsystem/SSshuttles.dm:329]:
```suggestion
					qdel(P, force = TRUE)
```

File: code/modules/shuttle/emergency.dm [hunk 3]
```diff
@@ -227,17 +228,17 @@
...
 	switch(mode)
 		if(SHUTTLE_ESCAPE)
-			dtime = max(SSshuttle.emergencyEscapeTime - dtime, 0)
+			dtime = max(SSshuttle.emergencyEscapeTime / shuttle_speed_factor - dtime, 0)
 		if(SHUTTLE_DOCKED)
 			dtime = max(SSshuttle.emergencyDockTime - dtime, 0)
 		else
 
-			dtime = max(SSshuttle.emergencyCallTime - dtime, 0)
+			dtime = max(SSshuttle.emergencyCallTime / shuttle_speed_factor - dtime, 0)
```
Comment (Sirryan2002) [inline code/modules/shuttle/emergency.dm:241]: use parentheses here to specify the order of operations clearly

File: code/modules/shuttle/shuttle.dm [hunk 4]
```diff
@@ -694,8 +698,8 @@
 	if(divisor <= 0)
 		divisor = 10
 	if(!timer)
-		return round(callTime/divisor, 1)
-	return max( round((timer+callTime-world.time)/divisor,1), 0 )
+		return round(callTime / shuttle_speed_factor / divisor, 1)
+	return max(round((timer + callTime - world.time) / divisor / shuttle_speed_factor, 1), 0)
```
Comment (Sirryan2002) [inline code/modules/shuttle/shuttle.dm:702]: same here

File: code/modules/supply/supply_packs/pack_abstract.dm [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+/datum/supply_packs/abstract/shuttle/narnar
+	contraband = TRUE
+	template = /datum/map_template/shuttle/emergency/narnar
+
+/datum/supply_packs/abstract/shuttle/old
+	template = /datum/map_template/shuttle/emergency/old
+
+/datum/supply_packs/abstract/shuttle/jungle
+	template = /datum/map_template/shuttle/emergency/jungle
```
Comment (Sirryan2002) [inline code/modules/supply/supply_packs/pack_abstract.dm:89]: minimum $2500 credits. 
-Anything with potentially hazardous stuff on-board or little room should be upwards $3500

File: code/__DEFINES/economy_defines.dm (hunk 1 above)
Comment (lewcc) [reply code/__DEFINES/economy_defines.dm:84]: I don't think this even gets used anymore tbh, all the orderable shuttles are now datumized

**Comment by DogeDogIs** [comment]
view contents could be a image of the shuttle 

or no just a ideia, IF that is possible

**Comment by spacegeese** [comment]
is there a shuttle that's literally just an escape pod?

**Comment by lewcc** [comment]
I won't be adding new shuttles until afterwards, just including existing ones for now.

As for an image preview, that might be a little iffy, though I appreciate the idea

File: code/modules/supply/supply_packs/pack_abstract.dm (hunk 5 above)
Comment (lewcc) [reply code/modules/supply/supply_packs/pack_abstract.dm]: Feel free to take another look. I added some base prices and made the clown shuttle cmag-only

File: code/controllers/subsystem/SSshuttles.dm [hunk 6]
```diff
@@ -303,4 +310,111 @@ SUBSYSTEM_DEF(shuttle)
...
+/datum/controller/subsystem/shuttle/proc/load_template(datum/map_template/shuttle/S)
+	// load shuttle template, centred at shuttle import landmark,
+	if(loading_shuttle_at_preview_template)
+		CRASH("A shuttle was already loading at the preview template when another was loaded")
+
+	loading_shuttle_at_preview_template = TRUE
+	var/turf/landmark_turf = get_turf(locate("landmark*Shuttle Import"))
+	S.load(landmark_turf, centered = TRUE)
+
+	var/affected = S.get_affected_turfs(landmark_turf, centered=TRUE)
```
Comment (Sirryan2002) [inline code/controllers/subsystem/SSshuttles.dm:323]:
```suggestion
	var/affected = S.get_affected_turfs(landmark_turf, centered = TRUE)
```

File: code/controllers/subsystem/SSshuttles.dm [hunk 7]
```diff
@@ -303,4 +310,111 @@ SUBSYSTEM_DEF(shuttle)
...
+
+	loaded_shuttle.register()
+
+	// TODO indicate to the user that success happened, rather than just
+	// blanking the modification tab
+
+	custom_escape_shuttle_loading = FALSE
+	return loaded_shuttle
+
+
```
Comment (DGamerL) [inline code/controllers/subsystem/SSshuttles.dm:419]:
```suggestion
```

File: code/modules/shuttle/shuttle.dm [hunk 8]
```diff
@@ -223,6 +223,10 @@
 	var/travelDir = 0				//direction the shuttle would travel in
 	var/rebuildable = 0				//can build new shuttle consoles for this one
 
+
```
Comment (DGamerL) [inline code/modules/shuttle/shuttle.dm:226]:
```suggestion
```

File: code/modules/shuttle/shuttle_manipulator.dm [hunk 9]
```diff
@@ -12,7 +12,7 @@
 	icon = 'icons/obj/machines/shuttle_manipulator.dmi'
 	icon_state = "holograph_on"
 	/// Used for cooldown, very obvious name, required due to shuttles spawning in the same location and causing the server to implode
-	var/shuttle_and_preview_cooldown = 0
+	var/shuttle_and_preview_cooldown = 0  // TODO MAKE SURE THIS NEW BEHAVIOR IS PRESENT
```
Comment (DGamerL) [inline code/modules/shuttle/shuttle_manipulator.dm:15]: Is this done? Or a TODO for a PR in the future?

File: code/modules/shuttle/shuttle_manipulator.dm [hunk 10]
```diff
@@ -200,16 +200,36 @@
...
+		preview_shuttle = SSshuttle.load_template(loading_template)
+		preview_template = loading_template
+
+	SSshuttle.replace_shuttle(preview_shuttle)
+
+	existing_shuttle = null
+	preview_shuttle = null
+	preview_template = null
+	selected = null
+
```
Comment (DGamerL) [inline code/modules/shuttle/shuttle_manipulator.dm:222]:
```suggestion
```

File: code/modules/supply/supply_packs/pack_shuttle.dm [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+		SP.special_enabled = FALSE
+	SSshuttle.custom_shuttle_ordered = TRUE
+	var/obj/docking_port/mobile/shuttle = SSshuttle.load_template(template)
+	shuttle.shuttle_speed_factor = speed_factor
+	SSshuttle.replace_shuttle(shuttle)
+
+/datum/supply_packs/abstract/admin_notify/donations
+	name = "Donation to Lonely Corgi Foundation"
+	manifest = "1% of every donation goes towards supporting corgis in need."
+	cost = 500
```
Comment (DGamerL) [inline code/modules/supply/supply_packs/pack_shuttle.dm:63]: Can you put this in the changelog?

File: code/modules/supply/supply_packs/pack_shuttle.dm [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+/datum/supply_packs/abstract/shuttle/old
+	cost = 2500
+	template = /datum/map_template/shuttle/emergency/old
+
+// this one isn't great but it isn't horrible either
+
+/datum/supply_packs/abstract/shuttle/cramped
+	cost = 3750
+	template = /datum/map_template/shuttle/emergency/cramped
+
```
Comment (DGamerL) [inline code/modules/supply/supply_packs/pack_shuttle.dm:84]:
```suggestion
```

File: code/modules/supply/supply_packs/pack_shuttle.dm [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+	// these are special, but start enabled. As soon as one is ordered, all variants of this should be disabled.
+	special = TRUE
+	special_enabled = TRUE
+	/// The shuttle that this pack corresponds to ordering.
+	/// If given on init, will fill out the pack automagically.
+	var/datum/map_template/shuttle/template
+	/// The base speed multiplier to apply to the shuttle. 2 means it arrives twice as fast.
+	var/speed_factor = 1
+
+
```
Comment (DGamerL) [inline code/modules/supply/supply_packs/pack_shuttle.dm:20]:
```suggestion
```

File: code/modules/supply/supply_packs/pack_shuttle.dm [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+	template = /datum/map_template/shuttle/emergency/cramped
+
+
+// these, otoh, have some pretty silly features, and are hidden behind emag
+
+/datum/supply_packs/abstract/shuttle/military
+	// if you can pool enough money together for this one, maybe you deserve it
+	speed_factor = 1.5
+	cost = 4250
+	template = /datum/map_template/shuttle/emergency/military
```
Comment (DGamerL) [inline code/modules/supply/supply_packs/pack_shuttle.dm:92]: Shouldn't this have the emag hidden variable? Or is the comment incorrect?

File: code/modules/supply/supply_packs/pack_shuttle.dm [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+	speed_factor = 0.75  // this one's a little slower, enjoy your ride!
+	cmag_hidden = TRUE
+	cost = 500  // let the clown have it
+	template = /datum/map_template/shuttle/emergency/clown
+
+/datum/supply_packs/abstract/shuttle/narnar
+	cost = 3000
+	hidden = TRUE
+	template = /datum/map_template/shuttle/emergency/narnar
+
```
Comment (DGamerL) [inline code/modules/supply/supply_packs/pack_shuttle.dm:104]:
```suggestion
```

**Review by DGamerL — COMMENTED** [review]
Code looks pretty good at a glance, it's mostly nitpicks and small oversights

File: code/modules/shuttle/shuttle_manipulator.dm (hunk 9 above)
Comment (lewcc) [reply code/modules/shuttle/shuttle_manipulator.dm:15]: Good catch. I need to verify that the manipulator still has the delay that it had before.

File: code/modules/supply/supply_packs/pack_shuttle.dm (hunk 14 above)
Comment (lewcc) [reply code/modules/supply/supply_packs/pack_shuttle.dm:90]: I kind of changed my mind for that one, in favor of making it very expensive

File: code/modules/supply/supply_packs/pack_shuttle.dm (hunk 11 above)
Comment (lewcc) [reply code/modules/supply/supply_packs/pack_shuttle.dm]: Ended up modifying this to be hidden, so admins can (with the right know-how) create new supply packs like this.

File: code/controllers/subsystem/SSshuttles.dm [hunk 16]
```diff
@@ -303,4 +310,110 @@ SUBSYSTEM_DEF(shuttle)
...
+		return port
+
+	var/msg = "load_template(): Shuttle Template [S.mappath] has no mobile docking port. Aborting import."
+
+	for(var/T in affected)
+		var/turf/T0 = T
+		T0.contents = null
+
+	message_admins(msg)
+	WARNING(msg)
```
Comment (DGamerL) [inline code/controllers/subsystem/SSshuttles.dm:363]: We don't change anything about the message here, right? Seems a bit weird to be split up from the messages

File: code/controllers/subsystem/SSshuttles.dm [hunk 17]
```diff
@@ -303,4 +310,110 @@ SUBSYSTEM_DEF(shuttle)
...
+		throw EXCEPTION(m)
+
+	var/result = loaded_shuttle.canDock(dock)
+	// truthy value means that it cannot dock for some reason
+	// but we can ignore the someone else docked error because we'll
+	// be moving into their place shortly
+	if((result != SHUTTLE_CAN_DOCK) && (result != SHUTTLE_SOMEONE_ELSE_DOCKED))
+
+		var/m = "Unsuccessful dock of [loaded_shuttle] ([result])."
+		message_admins("[m]")
```
Comment (DGamerL) [inline code/controllers/subsystem/SSshuttles.dm:397]:
```suggestion
		message_admins(m)
```

File: code/controllers/subsystem/SSshuttles.dm (hunk 16 above)
Comment (lewcc) [reply code/controllers/subsystem/SSshuttles.dm:363]: That's why we send it out to both with the same var, unless I'm misunderstanding you?

**Review by S34NW — APPROVED** [review]
No major issues over TM period, lets go

**Pull request merged**